src/OpenFeature/EventExecutor.cs

```diff
@@ -0,0 +1,368 @@
+            {
+                return true;
+            }
+            foreach (var providerReference in this._namedProviderReferences.Values)
+            {
+                if (providerReference == provider)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private bool IsProviderActive(FeatureProviderReference providerRef)
+        {
+            return this._activeSubscriptions.Contains(providerRef);
+        }
+
+        private void EmitOnRegistration(FeatureProviderReference provider, ProviderEventTypes eventType, EventHandlerDelegate handler)
+        {
+            if (provider == null)
+            {
+                return;
+            }
+            var status = provider.Provider.GetStatus();
+
+            var message = "";
+            if (status == ProviderStatus.Ready && eventType == ProviderEventTypes.ProviderReady)
+            {
+                message = "Provider is ready";
+            }
+            else if (status == ProviderStatus.Error && eventType == ProviderEventTypes.ProviderError)
+            {
+                message = "Provider is in error state";
+            }
+            else if (status == ProviderStatus.Stale && eventType == ProviderEventTypes.ProviderStale)
+            {
+                message = "Provider is in stale state";
+            }
+
+            if (message != "")
+            {
+                try
+                {
+                    handler.Invoke(new ProviderEventPayload
+                    {
+                        ProviderName = provider.Provider.GetMetadata().Name,
+                        Type = eventType,
+                        Message = message
+                    });
+                }
+                catch (Exception exc)
+                {
+                    Debug.WriteLine(exc);
+                }
+            }
+        }
+
+        private async void ProcessFeatureProviderEventsAsync(object providerRef)
+        {
+            while (true)
+            {
+                var typedProviderRef = (FeatureProviderReference)providerRef;
+                if (typedProviderRef.Provider.GetEventChannel() == null)
+                {
+                    return;
+                }
+                var item = await typedProviderRef.Provider.GetEventChannel().Reader.ReadAsync().ConfigureAwait(false);
+
+                switch (item)
+                {
+                    case ProviderEventPayload eventPayload:
+                        await this.EventChannel.Writer.WriteAsync(new Event { Provider = typedProviderRef, EventPayload = eventPayload }).ConfigureAwait(false);
+                        break;
+                    case ShutdownSignal _:
+                        typedProviderRef.ShutdownSemaphore.Release();
+                        return;
+                }
+            }
+        }
+
+        // Method to process events
+        private async void ProcessEventAsync()
```

Comment:
This tastes like Go (not that I have a problem with it, it's clear to me).

Reply:
Yes, I think most of the implementation logic here is quite similar to Go-sdk. And I think that's great as we can correlate implementation and compare them :)